src/main/scala/com/softwaremill/realworld/users/UsersRepository.scala

```diff
@@ -0,0 +1,69 @@
+package com.softwaremill.realworld.users
+
+import com.softwaremill.realworld.utils.Exceptions
+import io.getquill.*
+import zio.{IO, UIO, ZIO, ZLayer}
+
+import java.sql.SQLException
+import javax.sql.DataSource
+
+class UsersRepository(quill: SqliteZioJdbcContext[SnakeCase], dataSource: DataSource):
+
+  private val dsLayer: ZLayer[Any, Nothing, DataSource] = ZLayer.succeed(dataSource)
+
+  import quill.*
+
+  def findById(id: Int): IO[Exception, Option[UserData]] = run(for {
+    usr <- querySchema[UserSessionRow](entity = "users_sessions") if usr.userId == lift(id)
+    ur <- querySchema[UserRow](entity = "users") if ur.userId == lift(id)
+  } yield (ur, usr.token))
+    .map(_.headOption)
+    .map(_.map(user))
+    .provide(dsLayer)
+
+  def findByEmail(email: String): IO[Exception, Option[UserData]] = run(for {
+    ur <- querySchema[UserRow](entity = "users") if ur.email == lift(email)
+    usr <- querySchema[UserSessionRow](entity = "users_sessions") if usr.userId == ur.userId
+  } yield (ur, usr.token))
+    .map(_.headOption)
+    .map(_.map(user))
+    .provide(dsLayer)
+
+  def add(user: UserRegisterData): IO[Exception, UserData] = run(
+    quote(
+      query[UserRow]
+        .insertValue(
+          lift(
+            UserRow(
+              0,
+              user.email,
+              user.username,
+              user.password,
+              ""
+            )
+          )
+        )
+        .returningGenerated(_.userId)
+    )
+  ).flatMap(findById)
```

Comment:
perhaps this could be changed in way that User is not selected from DB (just built and returned using input data)

Reply:
if we have the id and the other data, than this should be possible, sure - just `.map` over the returned value